### DISCUSSION TITLE
Is the order of `AddTransforms` guaranteed?

### DISCUSSION THREAD

**Discussion by PieterBoeren** [dotnet/yarp#1822] [Q&A] [answered]
Probably a simple question, but couldn't really find it in the documentation:

Is the order of the `AddTransforms` guaranteed? So if I do something like this in my startup:

```C#
services
        .AddReverseProxy()
        .LoadFromConfig(Configuration.GetSection("ReverseProxy"))
        .AddTransforms<Transform1>()
        .AddTransforms<Transform2>()
        .AddTransforms<Transform3>();
```

Will the order always be:
1) Transforms in the config file
2) Transforms from `Transform1`
3) Transforms from `Transform2`
4) Transforms from `Transform3`

**Comment by MihaZupan** [comment, marked as answer]
Yes.

The only mention I see in the docs is under [`RequestTransforms`](https://microsoft.github.io/reverse-proxy/articles/transforms.html#request-transforms)
> These transforms are applied in the order they are specified in the route configuration.

We could be more explicit about ordering guarantees at the top of the article.

We build the list of transforms as follows:
1. Run all `ITransformFactory`es in order of registration for each raw transform (think config entries).
2. Run all `ITransformProvider`s in order of registration

When processing requests, we use the order that was set on the `TransformBuilderContext`.

**Reply by PieterBoeren** [reply]
Thanks!
